File: Лабораторна робота 4/lab4py/lab4py/func.py
import random
import math

class Number:
    def multiply(self, times):
        pass
    def divide(self, times):
        pass
    def findAbs(self):
        pass

class Rational(Number):
    def __init__(self, Re):
        self.real = Re
    def multiply(self, times):
        self.real *= times
    def divide(self, times):
        self.real /= times
    def findAbs(self):
        return math.sqrt(math.pow(self.real, 2))


class Complex(Number):
    def __init__(self, Re, Im):
        self.real = Re
        self.imaginary = Im
    def multiply(self, times):
        self.real *= times
        self.imaginary *= times
    def divide(self, times):
        self.real /= times
        self.imaginary /= times
    def findAbs(self):
        return math.sqrt(math.pow(self.real, 2) + math.pow(self.imaginary, 2))

def RatList(size):
    return [Rational(round(random.uniform(-100, 100), 3)) for i in range(size)]

def displayRat(Rat):
    for i in range(len(Rat)):
        x = round(Rat[i].real, 3)
        print(f"{i+1}) {x}")

def CompList(size):
    return [Complex(round(random.uniform(-100, 100), 3), round(random.uniform(-100, 100), 3)) for i in range(size)]

def displayComp(Comp):
    for i in range(len(Comp)):
        x = round(Comp[i].real, 3)
        y = round(Comp[i].imaginary, 3)
        print(f"{i+1}) {x} + {y}*i")

def changeRat(Rat):
    for i in range(len(Rat)):
        Rat[i].divide(2)

def changeComp(Comp):
    for i in range(len(Comp)):
        Comp[i].multiply(3)

def findSum(Rat, Comp):
    sum = 0
    for i in range(len(Rat)):
        sum += Rat[i].findAbs()
    for i in range(len(Comp)):
        sum += Comp[i].findAbs()
    return sum

File: Лабораторна робота 4/lab4cpp/lab4cpp/func.cpp
#include "func.h"

Number::Number() {}

Rational::Rational() {}

Rational::Rational(double Re) {
	real = Re;
}

void Rational::multiply(double times) {
	real *= times;
}

void Rational::divide(double times) {
	real /= times;
}

double Rational::getReal() {
	return real;
}

double Rational::findAbs() {
	return sqrt(real * real);
}


Complex::Complex() {}

Complex::Complex(double Re, double Im) {
	real = Re;
	imaginary = Im;
}

void Complex::multiply(double times) {
	real *= times;
	imaginary *= times;
}

void Complex::divide(double times) {
	real /= times;
	imaginary /= times;
}

double Complex::getReal() {
	return real;
}

double Complex::getImag() {
	return imaginary;
}

double Complex::findAbs() {
	return sqrt(real * real + imaginary * imaginary);
}

int input() {	
	int num = 0;
	do {
		std::cout << "enter the size: "; std::cin >> num;
	} while (num <= 0);
	return num;
}

Rational* RatArray(int size) {
	Rational* Array = new Rational[size];
	double min = -100.0;
	double max = 100.0;
	for (size_t i = 0; i < size; i++) {
		double Re = (double)(rand()) / RAND_MAX * (max - min) + min;
		Rational R(Re);
		Array[i] = R;
	}
	return Array;
}

Complex* CompArray(int size) {
	Complex* Array = new Complex[size];
	double min = -100.0;
	double max = 100.0;
	for (int i = 0; i < size; i++) {
		double Re = (double)(rand()) / RAND_MAX * (max - min) + min;
		double Im = (double)(rand()) / RAND_MAX * (max - min) + min;
		Complex C(Re, Im);
		Array[i] = C;
	}
	return Array;
}

void displayRat(Rational* Array, int size) {
	for (int i = 0; i < size; i++)
	{
		printf("%d) %3.3f \n", i + 1, Array[i].getReal());
	}
	std::cout << std::endl;
}

void displayComp(Complex* Array, int size) {
	for (int i = 0; i < size; i++)
	{
		printf("%d) %3.3f + %3.3f*i \n", i + 1, Array[i].getReal(), Array[i].getImag());
	}
	std::cout << std::endl;
}

void changeRat(Rational* Array, int size) {
	for (size_t i = 0; i < size; i++)
	{
		Array[i].divide(2);
	}
}

void changeComp(Complex* Array, int size) {
	for (size_t i = 0; i < size; i++)
	{
		Array[i].multiply(3);
	}
}

double findSum(Rational* Arr1, int m, Complex* Arr2, int n) {
	double sum = 0;
	for (size_t i = 0; i < m; i++)
	{
		sum += Arr1[i].findAbs();
	}
	for (size_t i = 0; i < n; i++)
	{
		sum += Arr2[i].findAbs();
	}
	return sum;
}